src/main/resources/common-services/CDAP/4.0.0/service_advisor.py

```diff
@@ -46,10 +49,90 @@ def colocateService(self, hostsComponentsMap, serviceComponents):
                     hostComponents.remove({"name": "CDAP_MASTER"})
 
     def getServiceConfigurationRecommendations(self, configurations, clusterSummary, services, hosts):
-        pass
+        # check Zookeeper configuration
+        if "zoo.cfg" in services["configurations"]:
+            zoo_cfg = services["configurations"]["zoo.cfg"]["properties"]
+            putZooCfgProperty = self.putProperty(configurations, "zoo.cfg", services)
+
+        for property, desired_value in self.getZooCfgDesiredValues().iteritems():
+            if property not in zoo_cfg or zoo_cfg[property] != desired_value:
+                putZooCfgProperty(property, desired_value)
+
+    def getZooCfgDesiredValues(self):
+        yarn_site_desired_values = {
+            "maxClientCnxns": "0"
+            }
+        return yarn_site_desired_values
 
     def getServiceComponentLayoutValidations(self, services, hosts):
         return []
 
     def getServiceConfigurationsValidationItems(self, configurations, recommendedDefaults, services, hosts):
+        # validate recommended properties in yarn-site
+        siteName = "yarn-site"
+        method = self.validateYarnSiteConfigurations
+        items = self.validateConfigurationsForSite(configurations, recommendedDefaults, services, hosts, siteName, method)
+
+        return items
+
+    def validateYarnSiteConfigurations(self, properties, recommendedDefaults, configurations, services, hosts):
+        yarn_site = properties
+        validationItems = []
+
+        # Calculate yarn available resources
+        nodeManagerHost = self.getHostsWithComponent("YARN", "NODEMANAGER", services, hosts)
+        nodeManagerCpu = int(len(nodeManagerHost)) * int(configurations["yarn-site"]["properties"]["yarn.nodemanager.resource.cpu-vcores"])
+        nodeManagerMem = int(len(nodeManagerHost)) * int(configurations["yarn-site"]["properties"]["yarn.nodemanager.resource.memory-mb"])
+
+        # calculate cdap resources
+        cdapProperties = configurations["cdap-site"]["properties"]
+        cdapCpu = int((int(cdapProperties["dataset.executor.container.num.cores"]) +
+                       int(cdapProperties["messaging.container.num.cores"]) +
+                       int(cdapProperties["stream.container.num.cores"]) +
+                       int(cdapProperties["metrics.num.cores"]) +
+                       int(cdapProperties["log.saver.container.num.cores"]) +
+                       int(cdapProperties["explore.executor.container.num.cores"]) +
+                       int(cdapProperties["metrics.processor.num.cores"]) +
+                       int(cdapProperties["data.tx.num.cores"]) +
+                       int(cdapProperties["master.service.num.cores"])))
```

Comment:
This is OK, for now. However, we'll want to change this to pull the list of property names from the configuration.

Something like (pseudo-code):
```
cdapCpu = 0
for name, value in cdapProperties.iteritems():
  if name match /*.num.cores/:
    cdapCpu += value
```

Otherwise, this code has to be updated whenever new container services are added/removed. Let's leave this, but file a JIRA to improve in 4.3's release.